README.md
# DESI Survey Simulations

This package simulates the nightly scheduling of observations during the DESI survey, using randomly generated observing conditions.

For a tutorial on installing the necessary software and running simulations, see the [tutorial](http://surveysim.readthedocs.io/en/latest/tutorial.html).


doc/changes.rst
====================
surveysim change log
====================

0.11.1 (unreleased)
-------------------

* No changes yet.

0.11.0 (2019-08-09)
-------------------

* Pass dummy sky level to desisurvey scheduler.next_tile; needed to match
  API change in desisurvey PR #99. (surveysim PR `#64`_).
  Requires desisurvey 0.12.0 or later.

.. _`#64`: https://github.com/desihub/surveysim/pull/64

0.10.1 (2018-12-16)
-------------------

* Include EXTNAME in output files (PR `#63`_).

.. _`#63`: https://github.com/desihub/surveysim/pull/63

0.10.0 (2018-11-26)
-------------------

This version is a major refactoring of the code to take advantage of the
refactoring in desisurvey 0.11.0 and simplify the simulation classes
used to track survey statistics and exposure metadata (PR `#60`_).

* Add new modules: exposures, stats.
* Add new simulate_night that supercedes the original nightOps.
* Use new Scheduler and ExposureTimeCalculator from desisurvey.
* Requires desisurvey 0.11.0.
* Refactor desisurvey.ephemerides -> desisurvey.ephem and use get_ephem().
* Update the tutorial.

.. _`#60`: https://github.com/desihub/surveysim/pull/60


0.9.2 (2018-10-02)
------------------

* Replay historical Mayall daily weather.
* Implement partial-night dome closure.
* Requires desimodel >= 0.9.8 and desisurvey >= 0.10.4.

0.9.1 (2018-06-27)
------------------

* Do arc exposures before flat exposures (PR `#57`_).

.. _`#57`: https://github.com/desihub/surveysim/pull/57

0.9.0 (2017-11-09)
------------------

* Add ``surveysim.util.add_calibration_exposures()``, to add simulated
  calibration exposures to a set of science exposures (PR `#55`_).

.. _`#55`: https://github.com/desihub/surveysim/pull/55

0.8.2 (2017-10-09)
------------------

* Use new desisurvey config api (requires desisurvey >= 0.9.3)
* Add support for optional depth-first survey strategy.
* Docs now auto-generated at http://surveysim.readthedocs.io/en/latest/

0.8.1 (2017-09-20)
------------------

* Adds surveysim --config-file option (PR `#49`_); requires desisurvey/0.9.1.

.. _`#49`: https://github.com/desihub/surveysim/pull/49

0.8.0 (2017-09-11)
------------------

* Track API changes in desisurvey 0.9.0.
* The surveysim script is now called once per night, alternating with a
  surveyplan script that lives in the desisurvey package.
* See https://www.youtube.com/watch?v=vO1QZD_aCIo for a visualization of the
  full 5-year survey simulation that matches DESI-doc-1767-v3.

0.7.1 (2017-08-07)
------------------

* Use new desimodel.weather to randomly sample seeing and transparency.
  Requires desimodel >= 0.8.0.

0.7.0 (2017-06-18)
------------------

* First implementation of fiber-assignment groups and priorities.
* Integration with the new desisurvey surveyplan script.
* Create tutorial document and sample automation scripts.

0.6.0 (2017-06-05)
------------------

* Add strategy, weights options to surveysim script.
* Add hooks for using greedy scheduler
* Terminate exposures at sunset

0.5.0 (2017-05-10)
------------------

* Use desisurvey.config to manage all non-simulation configuration data.
* Unify different output files with overlapping contents into single output
  managed by desisurvey.progress.
* Overhaul of weather simulator to generate continuous stationary time series
  that are independent of the observing sequence.  Use desimodel.seeing.
* Simulate multiple exposures for cosmics and more realistic overhead.
* Clean up of README, docstrings, imports, unit tests, requirements, unused code.

0.4.1 (2017-04-13)
------------------

* Fixed package names to work with desisurvey >= 0.4.0

0.4.0 (2017-04-04)
------------------

* Adds unit tests
* removes data/tile-info.fits (not used here; was moved to desisurvey)
* adds nightops.py (from desisurvey, used here but not there)
* create surveysim command-line script
* use new desisurvey config machinery (first steps, in progress)

0.3.1 (2016-12-21)
------------------

* Fixed outlier HA tile assignments around RA 200-220 (PR #26)
* Added 7 day shutdown around full moon (PR #25)

0.3.0 (2016-11-29)
------------------

* Moved non-simulation specific parts to desisurvey

0.2.0 (2016-11-18)
------------------

* Modified some file names
* Moved some functions from one file to another

0.1.1 (2016-11-14)
------------------

* fixed crash at end and data/ install (PR #3)
* initial tests for NERSC install

0.1.0 and prior
---------------

* No changes.rst yet


calendar.py
from astropy.time import Time, TimezoneInfo
from datetime import datetime
import astropy.units as u
import numpy as np

"""
Line format
0, 1, 2: evening day, month, date
3: /
4, 5, 6: morning day, month, date
7: MJD of local midnight
8, 9, 10: LMST of midnight hours, minutes, seconds
11, 12: Sun set hours, minutes
13, 14: evening twilight end hours, minutes
15, 16: morning twilight begin hours, minutes
17, 18: Sun rise hours, minutes
19, 20: evening twilight LST hours, minutes
21, 22: morning twilight LST hours, minutes
23, 24: Moon rise hours, minutes                   23: ....                                  23, 24: Moon rise hours, minutes
25, 26: Moon set hours, minutes                    24, 25: Moon set hours, minutes           25: ....
27: Moon illumination fraction                     26: Moon illumination fraction            26: Moon illumination fraction
28, 29: Moon RA hours, minutes                     27, 28: Moon RA hours, minutes            27, 28: Moon RA hours, minutes
30: -
32, 33: Moon DEC degrees, minutes                  31, 32: Moon DEC degrees, minutes         31, 32: Moon DEC degrees, minutes
"""
daymonth = np.array([31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]) # This is only valid for leap years like 2016.

def monthNum(monthname):
    month = 0
    if monthname == 'Jan': month = 1
    elif monthname == 'Feb': month = 2
    elif monthname == 'Mar': month = 3
    elif monthname == 'Apr': month = 4
    elif monthname == 'May': month = 5
    elif monthname == 'Jun': month = 6
    elif monthname == 'Jul': month = 7
    elif monthname == 'Aug': month = 8
    elif monthname == 'Sep': month = 9
    elif monthname == 'Oct': month = 10
    elif monthname == 'Nov': month = 11
    elif monthname == 'Dec': month = 12
    return month

def obsCalendar(calfile):
    utc_minus7 = TimezoneInfo(utc_offset=-7*u.hour)
    year = 2016 # Testing calendar with only one year
    cal = []
    with open(calfile,'r') as f:
        while True:
            line = f.readline()
            if not line: break
            if line[0] == '#': continue
            entries = line.split()
            month = monthNum(entries[1])
            day = int(entries[2])
            MJDmidnight = float(entries[7])
            sunset_h = int(entries[11])
            sunset_m = int(entries[12])
            Tsunset = Time(datetime(year, month, day, sunset_h, sunset_m, 0, tzinfo=utc_minus7))
            etwi_h = int(entries[13])
            etwi_m = int(entries[14])
            Tetwi = Time(datetime(year, month, day, etwi_h, etwi_m, 0, tzinfo=utc_minus7))
            mtwi_h = int(entries[15])
            mtwi_m = int(entries[16])
            yearnew = year
            monthnew = month
            daynew = day + 1
            if daynew > daymonth[month-1]:
                daynew = 1
                monthnew = month + 1
                if monthnew > 12:
                    monthnew = 1
                    yearnew = year +1
            Tmtwi = Time(datetime(yearnew, monthnew, daynew, mtwi_h, mtwi_m, 0, tzinfo=utc_minus7))
            sunrise_h = int(entries[17])
            sunrise_m = int(entries[18])
            Tsunrise = Time(datetime(yearnew, monthnew, daynew, sunrise_h, sunrise_m, 0, tzinfo=utc_minus7))
            if entries[23] == ".....":
                MJDmoonrise = -1.0
                moonset_h = int(entries[24])
                moonset_m = int(entries[25])
                yearnew = year
                monthnew = month
                daynew = day
                if moonset_h < 12:
                    daynew = day + 1
                    if daynew > daymonth[month-1]:
                        daynew = 1
                        monthnew = month + 1
                        if monthnew > 12:
                            monthnew = 1
                            yearnew = year +1
                MJDmoonset = (Time(datetime(yearnew, monthnew, daynew, moonset_h, moonset_m, 0, tzinfo=utc_minus7))).mjd
            else:
                moonrise_h = int(entries[23])
                moonrise_m = int(entries[24])
                yearnew = year
                monthnew = month
                daynew = day
                if moonrise_h < 12:
                    daynew = day + 1
                    if daynew > daymonth[month-1]:
                        daynew = 1
                        monthnew = month + 1
                        if monthnew > 12:
                            monthnew = 1
                            yearnew = year +1
                MJDmoonrise = (Time(datetime(yearnew, monthnew, daynew, moonrise_h, moonrise_m, 0, tzinfo=utc_minus7))).mjd
                if entries[25] == ".....":
                    MJDmoonset = -1.0
                else:
                    moonset_h = int(entries[25])
                    moonset_m = int(entries[26])
                    yearnew = year
                    monthnew = month
                    daynew = day
                    if moonset_h < 12:
                        daynew = day + 1
                        if daynew > daymonth[month-1]:
                            daynew = 1
                            monthnew = month + 1
                            if monthnew > 12:
                                monthnew = 1
                                yearnew = year +1
                    MJDmoonset = (Time(datetime(yearnew, monthnew, daynew, moonset_h, moonset_m, 0, tzinfo=utc_minus7))).mjd
            if entries[23] == "....." or entries[25] == ".....":
                moonIllum = float(entries[26])
                moonRA = float(entries[27]) + float(entries[28])/60.0
                moonRA = moonRA * 15.0
                if entries[29] == "-":
                    moonDEC = -1.0 * (float(entries[30]) + float(entries[31])/100.0)
                else:
                    moonDec = float(entries[29]) + float(entries[30])/100.0
            else:
                moonIllum = float(entries[27])
                moonRA = float(entries[28]) + float(entries[29])/60.0
                if entries[30] == "-":
                    moonDEC = -1.0 * (float(entries[31]) + float(entries[32])/100.0)
                else:
                    moonDec = float(entries[30]) + float(entries[31])/100.0
            cal.append( {'MJDmidnight': MJDmidnight,
                         'MJDsunset': Tsunset.mjd,
                         'MJDsunrise': Tsunrise.mjd,
                         'MJDetwi': Tetwi.mjd,
                         'MJDmtwi': Tmtwi.mjd,
                         'MJDmoonrise': MJDmoonrise,
                         'MJDmoonset': MJDmoonset,
                         'MoonFrac': moonIllum,
                         'MoonRA': moonRA,
                         'MoonDEC': moonDEC} )
    return cal
